[WebGPU] Depth buffer / texture support

mwyrzykowski opened the pull request:
#### c578df6c965c6b8937a9b7d31de11a848fe2c8cc
<pre>
[WebGPU] Depth buffer / texture support
<a href="https://bugs.webkit.org/show_bug.cgi?id=249037">https://bugs.webkit.org/show_bug.cgi?id=249037</a>
&lt;radar://101571337&gt;

Reviewed by Dean Jackson.

Add depth buffer support along with a test case which shows
several intersecting rotating textured cubes.

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::validateRenderPassDescriptor const):
(WebGPU::CommandEncoder::beginRenderPass):
Create the depth state.

* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::setPipeline):
Set the depth stencil state if it exists.

* Source/WebGPU/WebGPU/RenderPipeline.h:
(WebGPU::RenderPipeline::create):
(WebGPU::RenderPipeline::depthStencilState const):

* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::topologyType):
Set input topology type.

(WebGPU::Device::validateRenderPipeline):
Depth stencil is supported.

(WebGPU::convertToMTLCompare):
Convert compare function to metal compare function.

(WebGPU::Device::createRenderPipeline):
(WebGPU::RenderPipeline::RenderPipeline):

* Websites/webkit.org/demos/webgpu/instanced-textured-cube.html: Added.
* Websites/webkit.org/demos/webgpu/scripts/instanced-textured-cube.js: Added.
(async helloCube.frameUpdate):
(async helloCube):
Add a test case showing 9 cubes rotating with different offsets and intersecting
other cubes to illustrate a working depth buffer.

Canonical link: <a href="https://commits.webkit.org/257807@main">https://commits.webkit.org/257807@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ef09be2723bb84213b7056b887b9a6e95846fd1a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99991 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9160 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33071 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109338 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10044 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86472 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92438 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107238 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105759 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/7589 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90872 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34306 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89495 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22263 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2955 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23776 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2915 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/46150 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9048 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43254 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5359 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4763 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

grorg commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -184,7 +180,24 @@ static MTLStoreAction storeAction(WGPUStoreOp storeOp)
-    // FIXME: Depth + stencil texture support
+    if (const auto* attachment = descriptor.depthStencilAttachment) {
+        const auto& mtlAttachment = mtlDescriptor.depthAttachment;
+        ASSERT("FIXME: Add support for depthReadOnly" && !attachment->depthReadOnly);
```
Can't believe I've never thought to use ASSERTs like this (with a message)!

mwyrzykowski replied:
oh I could have used `ASSERT_WITH_MESSAGE(` instead too, that might be more clear

mwyrzykowski committed:
[WebGPU] Depth buffer / texture support
https://bugs.webkit.org/show_bug.cgi?id=249037
<radar://101571337>

Reviewed by Dean Jackson.

Add depth buffer support along with a test case which shows
several intersecting rotating textured cubes.

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::validateRenderPassDescriptor const):
(WebGPU::CommandEncoder::beginRenderPass):
Create the depth state.

* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::setPipeline):
Set the depth stencil state if it exists.

* Source/WebGPU/WebGPU/RenderPipeline.h:
(WebGPU::RenderPipeline::create):
(WebGPU::RenderPipeline::depthStencilState const):

* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::topologyType):
Set input topology type.

(WebGPU::Device::validateRenderPipeline):
Depth stencil is supported.

(WebGPU::convertToMTLCompare):
Convert compare function to metal compare function.

(WebGPU::Device::createRenderPipeline):
(WebGPU::RenderPipeline::RenderPipeline):

* Websites/webkit.org/demos/webgpu/instanced-textured-cube.html: Added.
* Websites/webkit.org/demos/webgpu/scripts/instanced-textured-cube.js: Added.
(async helloCube.frameUpdate):
(async helloCube):
Add a test case showing 9 cubes rotating with different offsets and intersecting
other cubes to illustrate a working depth buffer.

Canonical link: https://commits.webkit.org/257807@main
